File: backend/core/skills/maintenance.py
"""Repository and sync health checks plus conservative repair actions."""
from __future__ import annotations

import logging
import os
import time
import uuid
from pathlib import Path
from typing import Any, Optional

from core.utils.content_hash import hash_dir
from core.utils.path_safety import is_path_within, norm_path, require_path_within
from core.db.store import SkillRecord, SkillStore, SkillTargetRecord, now_ms
from core.repo.community import resolve_community_repo_path, resolve_custom_repo_path
from core.skills.installer import is_skill_dir, parse_skill_md
from core.skills.install_service import build_skill_record
from core.skills.source_paths import normalize_source_type, resolve_skill_source_path
from core.skills.sync_engine import (
    _is_junction,
    _remove_path_any,
    sync_dir_copy_with_overwrite,
    sync_dir_for_tool_with_overwrite,
)
from core.tools.adapters import (
    adapter_by_key,
    effective_tool_adapters,
    is_tool_installed,
    resolve_default_path,
    resolve_project_path,
    scan_tool_dir,
)

logger = logging.getLogger(__name__)


def _target_base(target: SkillTargetRecord) -> Optional[str]:
    adapter = adapter_by_key(target.tool)
    if not adapter:
        return None
    if (target.scope or "global") == "project":
        if not target.project_path:
            return None
        return resolve_project_path(adapter, target.project_path)
    return resolve_default_path(adapter)


def _issue(
    code: str,
    severity: str,
    message: str,
    repair: Optional[str] = None,
    **data: Any,
) -> dict[str, Any]:
    payload = {
        "code": code,
        "severity": severity,
        "message": message,
        "repair_action": repair,
    }
    payload.update(data)
    return payload


def scan_sync_health(store: SkillStore) -> dict[str, Any]:
    """Scan DB, community repo and installed tool dirs without mutating state."""
    community_base = resolve_community_repo_path(store.db_path)
    custom_base = resolve_custom_repo_path(store.db_path)
    skills = store.list_skills()
    issues: list[dict[str, Any]] = []
    target_paths: set[str] = set()
    community_paths: set[str] = set()
    custom_paths: set[str] = set()

    for skill in skills:
        source_type = normalize_source_type(skill.source_type)
        source_label = "自制源目录" if source_type == "custom" else "中央仓库"
        try:
            source_path = resolve_skill_source_path(skill, store)
        except ValueError:
            issues.append(_issue(
                "source_path_outside_repo",
                "error",
                f"{source_label}路径超出允许范围 / source path escapes repo",
                skill_id=skill.id,
                skill_name=skill.name,
                community_path=skill.community_path,
            ))
            continue
        if source_type == "community":
            community_paths.add(norm_path(source_path))
        elif source_type == "custom":
            custom_paths.add(norm_path(source_path))
        if not source_path.is_dir():
            missing_code = "missing_source_dir" if source_type == "custom" else "missing_community_dir"
            issues.append(_issue(
                missing_code,
                "error",
                f"{source_label}中的 Skill 目录丢失 / source skill directory is missing",
                "mark_skill_missing",
                skill_id=skill.id,
                skill_name=skill.name,
                community_path=str(source_path),
            ))
            continue
        try:
            source_hash = hash_dir(source_path)
            if skill.content_hash and source_hash != skill.content_hash:
                issues.append(_issue(
                    "source_hash_drift",
                    "info",
                    f"{source_label}内容与数据库哈希不一致 / source content hash differs from DB",
                    "update_content_hash",
                    skill_id=skill.id,
                    skill_name=skill.name,
                    community_path=str(source_path),
                    stored_hash=skill.content_hash,
                    actual_hash=source_hash,
                ))
        except Exception as e:
            issues.append(_issue(
                "source_hash_failed",
                "info",
                f"无法计算{source_label}哈希 / cannot hash source skill",
                skill_id=skill.id,
                skill_name=skill.name,
                community_path=str(source_path),
                error=str(e),
            ))

        targets = store.list_skill_targets(skill.id)
        for target in targets:
            target_paths.add(norm_path(target.target_path))
            base = _target_base(target)
            if not base:
                issues.append(_issue(
                    "invalid_target_record",
                    "error",
                    "同步目标记录无法解析工具或项目路径 / invalid target record",
                    "delete_target_record",
                    skill_id=skill.id,
                    tool=target.tool,
                    target_path=target.target_path,
                ))
                continue
            if not is_path_within(target.target_path, base):
                issues.append(_issue(
                    "target_path_outside_tool_dir",
                    "error",
                    "同步目标路径超出工具目录 / target path escapes tool directory",
                    skill_id=skill.id,
                    tool=target.tool,
                    target_path=target.target_path,
                    expected_base=base,
                ))
                continue

            path = Path(target.target_path)
            exists = path.exists() or path.is_symlink() or _is_junction(path)
            if not exists:
                issues.append(_issue(
                    "missing_target_path",
                    "warning",
                    "数据库记录存在，但工具目录中的目标已丢失 / target path is missing",
                    "resync_target",
                    skill_id=skill.id,
                    tool=target.tool,
                    scope=target.scope,
                    project_path=target.project_path,
                    target_path=target.target_path,
                ))
                continue

            is_link = path.is_symlink() or _is_junction(path)
            if target.mode in {"symlink", "junction"} and not is_link:
                issues.append(_issue(
                    "mode_mismatch",
                    "warning",
                    "数据库记录为链接模式，但目标不是链接 / target mode does not match filesystem",
                    skill_id=skill.id,
                    tool=target.tool,
                    target_path=target.target_path,
                    mode=target.mode,
                ))

            if is_link and not Path(os.path.realpath(path)).exists():
                issues.append(_issue(
                    "broken_link",
                    "warning",
                    "同步链接已断开 / sync link target is missing",
                    "resync_target",
                    skill_id=skill.id,
                    tool=target.tool,
                    target_path=target.target_path,
                ))
                continue

            if is_link and norm_path(os.path.realpath(path)) != norm_path(source_path):
                issues.append(_issue(
                    "wrong_link_target",
                    "warning",
                    "同步链接指向的目录不是当前源路径 / link points to the wrong target",
                    skill_id=skill.id,
                    tool=target.tool,
                    target_path=target.target_path,
                    expected_path=str(source_path),
                    actual_path=os.path.realpath(path),
                ))
                continue

            if target.mode == "copy" and path.is_dir():
                try:
                    source_hash = hash_dir(source_path)
                    # 优先使用存储的 target_content_hash 快速比较，避免重新哈希
                    if target.target_content_hash and target.target_content_hash == source_hash:
                        # 存储的哈希与源一致，无需重新扫描
                        pass
                    else:
                        target_hash = hash_dir(path)
                        if source_hash != target_hash:
                            issues.append(_issue(
                                "stale_copy_target",
                                "warning",
                                "copy 模式目标内容落后于中央仓库 / copied target is stale",
                                "resync_copy_target",
                                skill_id=skill.id,
                                tool=target.tool,
                                target_path=target.target_path,
                                source_hash=source_hash,
                                target_hash=target_hash,
                            ))
                except Exception as e:
                    issues.append(_issue(
                        "target_hash_failed",
                        "info",
                        "无法计算同步目标哈希 / cannot hash target",
                        skill_id=skill.id,
                        tool=target.tool,
                        target_path=target.target_path,
                        error=str(e),
                    ))

    if community_base.is_dir():
        for item in sorted(community_base.iterdir()):
            if not item.is_dir() or item.name == ".snapshots":
                continue
            if not is_skill_dir(item):
                continue
            if norm_path(item) not in community_paths:
                issues.append(_issue(
                    "community_orphan_dir",
                    "info",
                    "Community Repo 中存在未登记的 Skill / community skill is not registered",
                    "register_community_skill",
                    community_path=str(item),
                    skill_name=item.name,
                ))

    if custom_base.is_dir():
        for item in sorted(custom_base.iterdir()):
            if not item.is_dir() or item.name == ".snapshots":
                continue
            if not is_skill_dir(item) and not item.name.endswith("-skills"):
                continue
            if norm_path(item) not in custom_paths:
                # 跳过已注册套件根目录下的子 skill
                if any(is_path_within(item, Path(cp)) for cp in custom_paths):
                    continue
                issues.append(_issue(
                    "custom_orphan_dir",
                    "info",
                    "自制仓库中存在未登记的 Skill / custom skill is not registered",
                    "register_custom_skill",
                    community_path=str(item),
                    skill_name=item.name,
                ))

    for adapter in effective_tool_adapters():
        if not is_tool_installed(adapter):
            continue
        skills_dir = resolve_default_path(adapter)
        for detected in scan_tool_dir(adapter, skills_dir):
            if norm_path(detected.path) in target_paths:
                continue
            if is_path_within(detected.path, community_base):
                continue
            if is_path_within(detected.path, custom_base):
                continue
            issues.append(_issue(
                "unmanaged_tool_skill",
                "info",
                "工具目录中存在未托管 Skill / unmanaged skill in tool directory",
                "import_tool_skill",
                tool=adapter.id.as_key(),
                skill_name=detected.name,
                target_path=detected.path,
                is_link=detected.is_link,
                link_target=detected.link_target,
            ))

    hashes: dict[str, list[SkillRecord]] = {}
    names: dict[str, list[SkillRecord]] = {}
    for skill in skills:
        if skill.content_hash:
            hashes.setdefault(skill.content_hash, []).append(skill)
        names.setdefault(skill.name.lower(), []).append(skill)
    for content_hash, group in hashes.items():
        if len(group) > 1:
            issues.append(_issue(
                "duplicate_content_hash",
                "info",
                "多个 Skill 内容完全相同 / duplicate skill content",
                skill_ids=[s.id for s in group],
                names=[s.name for s in group],
                content_hash=content_hash,
            ))
    for name, group in names.items():
        unique_hashes = {s.content_hash for s in group if s.content_hash}
        if len(group) > 1 and len(unique_hashes) > 1:
            issues.append(_issue(
                "same_name_different_content",
                "info",
                "多个同名 Skill 内容不同 / same name with different content",
                skill_ids=[s.id for s in group],
                name=name,
            ))

    return {
        "community_repo": str(community_base),
        "skills_checked": len(skills),
        "issues": issues,
        "summary": _summarize_issues(issues),
        "generated_at": int(time.time() * 1000),
    }


def _summarize_issues(issues: list[dict[str, Any]]) -> dict[str, int]:
    summary = {"error": 0, "warning": 0, "info": 0, "repairable": 0}
    for issue in issues:
        severity = issue.get("severity", "info")
        summary[severity] = summary.get(severity, 0) + 1
        if issue.get("repair_action"):
            summary["repairable"] += 1
    return summary


def repair_sync_health(store: SkillStore, dry_run: bool = True) -> dict[str, Any]:
    """Run conservative repairs. Destructive deletes are limited to DB records or managed targets."""
    report = scan_sync_health(store)
    operations: list[dict[str, Any]] = []

    for issue in report["issues"]:
        action = issue.get("repair_action")
        if action in {"register_community_skill", "register_custom_skill"}:
            operations.append({"action": action, "community_path": issue["community_path"]})
            if not dry_run:
                source_type = "custom" if action == "register_custom_skill" else "community"
                record = create_import_record_from_community(store, issue["community_path"], source_type)
                store.upsert_skill(record)
        elif action == "mark_skill_missing":
            operations.append({"action": action, "skill_id": issue["skill_id"]})
            if not dry_run:
                skill = store.get_skill_by_id(issue["skill_id"])
                if skill:
                    skill.status = "missing"
                    skill.updated_at = now_ms()
                    store.upsert_skill(skill)
        elif action == "update_content_hash":
            operations.append({"action": action, "skill_id": issue["skill_id"]})
            if not dry_run:
                skill = store.get_skill_by_id(issue["skill_id"])
                if skill:
                    skill.content_hash = issue.get("actual_hash")
                    skill.updated_at = now_ms()
                    store.upsert_skill(skill)
        elif action in {"delete_target_record", "resync_target", "resync_copy_target"}:
            operations.append({
                "action": action,
                "skill_id": issue.get("skill_id"),
                "tool": issue.get("tool"),
                "target_path": issue.get("target_path"),
            })
            if not dry_run and issue.get("skill_id") and issue.get("tool"):
                _repair_target_issue(store, issue)

    logger.info(
        "sync health repair %s: %d operations",
        "executed" if not dry_run else "planned (dry-run)",
        len(operations),
    )
    return {
        "dry_run": dry_run,
        "operations": operations,
        "operation_count": len(operations),
        "before": report,
        "after": scan_sync_health(store) if not dry_run else None,
    }


def _repair_target_issue(store: SkillStore, issue: dict[str, Any]) -> None:
    skill_id = issue["skill_id"]
    tool = issue["tool"]
    skill = store.get_skill_by_id(skill_id)
    if not skill:
        return
    try:
        source_path = resolve_skill_source_path(skill, store)
    except ValueError:
        return
    if not source_path.is_dir():
        return

    matching = [t for t in store.list_skill_targets(skill_id) if t.tool == tool]
    for target in matching:
        if issue.get("target_path") and norm_path(target.target_path) != norm_path(issue["target_path"]):
            continue
        base = _target_base(target)
        if not base:
            store.delete_skill_target(target.skill_id, target.tool, target.scope, target.project_path)
            continue
        target_path = require_path_within(target.target_path, base, "target path")
        if issue["repair_action"] == "delete_target_record":
            store.delete_skill_target(target.skill_id, target.tool, target.scope, target.project_path)
            continue
        _remove_path_any(target_path)
        if target.mode == "copy":
            result = sync_dir_copy_with_overwrite(source_path, target_path, overwrite=True)
        else:
            result = sync_dir_for_tool_with_overwrite(target.tool, str(source_path), target_path, overwrite=True)
        target.mode = result.mode_used.value
        target.status = "ok"
        target.last_error = None
        target.synced_at = now_ms()
        store.upsert_skill_target(target)
        logger.info("repair resynced target: skill=%s tool=%s mode=%s", skill_id, target.tool, target.mode)


def create_import_record_from_community(
    store: SkillStore, community_path: str | Path, source_type: str = "community"
) -> SkillRecord:
    path = Path(community_path)
    fm = parse_skill_md(path)
    content_hash = None
    try:
        content_hash = hash_dir(path)
    except Exception:
        pass
    return build_skill_record(
        skill_id=str(uuid.uuid4()),
        name=fm.name,
        description=fm.description,
        frontmatter=fm,
        source_type=source_type,
        source_ref=str(path),
        community_path=str(path),
        content_hash=content_hash,
    )

File: frontend/src-tauri/src/services/maintenance.rs
//! Maintenance service — mirrors `backend/core/skills/maintenance.py`.
//!
//! Provides sync health scanning, repair operations, and cache cleanup.

use serde::Serialize;
use std::collections::HashMap;
use std::path::Path;

use crate::db::{now_ms, Database};
use crate::models::Skill;
use crate::repo::community::{resolve_community_repo_path, resolve_custom_repo_path};
use crate::repo::scanner::is_skill_dir;
use crate::utils::content_hash;
use crate::utils::path_safety;

/// A single issue found during health scan.
#[derive(Debug, Clone, Serialize)]
pub struct HealthIssue {
    pub code: String,
    pub severity: String,
    pub message: String,
    pub repair_action: Option<String>,
    #[serde(flatten)]
    pub extra: HashMap<String, serde_json::Value>,
}

/// Summary of issues by severity.
#[derive(Debug, Clone, Serialize)]
pub struct IssueSummary {
    pub error: usize,
    pub warning: usize,
    pub info: usize,
    pub repairable: usize,
}

/// Full sync health report.
#[derive(Debug, Clone, Serialize)]
pub struct SyncHealthReport {
    pub community_repo: String,
    pub skills_checked: usize,
    pub issues: Vec<HealthIssue>,
    pub summary: IssueSummary,
    pub generated_at: i64,
}

/// Result of a repair operation.
#[derive(Debug, Clone, Serialize)]
pub struct RepairReport {
    pub dry_run: bool,
    pub operations: Vec<RepairOperation>,
    pub operation_count: usize,
    pub before: SyncHealthReport,
    pub after: Option<SyncHealthReport>,
}

#[derive(Debug, Clone, Serialize)]
pub struct RepairOperation {
    pub action: String,
    #[serde(flatten)]
    pub details: HashMap<String, serde_json::Value>,
}

fn issue(
    code: &str,
    severity: &str,
    message: &str,
    repair: Option<&str>,
) -> (HealthIssue, HashMap<String, serde_json::Value>) {
    (
        HealthIssue {
            code: code.to_string(),
            severity: severity.to_string(),
            message: message.to_string(),
            repair_action: repair.map(|r| r.to_string()),
            extra: HashMap::new(),
        },
        HashMap::new(),
    )
}

fn add_extra(issue: &mut HealthIssue, key: &str, value: serde_json::Value) {
    issue.extra.insert(key.to_string(), value);
}

fn summarize_issues(issues: &[HealthIssue]) -> IssueSummary {
    let mut summary = IssueSummary {
        error: 0,
        warning: 0,
        info: 0,
        repairable: 0,
    };
    for issue in issues {
        match issue.severity.as_str() {
            "error" => summary.error += 1,
            "warning" => summary.warning += 1,
            _ => summary.info += 1,
        }
        if issue.repair_action.is_some() {
            summary.repairable += 1;
        }
    }
    summary
}

/// Scan sync health without mutating state.
pub fn scan_sync_health(db: &Database) -> SyncHealthReport {
    use crate::repositories::{SkillTargetsRepository, SkillsRepository};

    let community_base = resolve_community_repo_path(db);
    let _custom_base = resolve_custom_repo_path(db);
    let skills_repo = SkillsRepository::new(db);
    let targets_repo = SkillTargetsRepository::new(db);

    let skills = skills_repo.list("manual").unwrap_or_default();
    let mut issues: Vec<HealthIssue> = Vec::new();
    let mut target_paths: std::collections::HashSet<String> = std::collections::HashSet::new();
    let mut community_paths: std::collections::HashSet<String> = std::collections::HashSet::new();
    let mut custom_paths: std::collections::HashSet<String> = std::collections::HashSet::new();

    for skill in &skills {
        let source_type = crate::repo::scanner::normalize_source_type(&skill.source_type);
        let source_label = if source_type == "custom" {
            "自制源目录"
        } else {
            "中央仓库"
        };

        // Check source path validity
        let source_path = resolve_skill_source_path(skill, db);
        let source_path = match source_path {
            Ok(p) => p,
            Err(_) => {
                let (mut iss, _) = issue(
                    "source_path_outside_repo",
                    "error",
                    &format!(
                        "{}路径超出允许范围 / source path escapes repo",
                        source_label
                    ),
                    None,
                );
                add_extra(&mut iss, "skill_id", serde_json::json!(skill.id));
                add_extra(&mut iss, "skill_name", serde_json::json!(skill.name));
                issues.push(iss);
                continue;
            }
        };

        if source_type == "community" {
            community_paths.insert(path_safety::norm_path(&source_path));
        } else {
            custom_paths.insert(path_safety::norm_path(&source_path));
        }

        // Check source directory exists
        if !source_path.is_dir() {
            let missing_code = if source_type == "custom" {
                "missing_source_dir"
            } else {
                "missing_community_dir"
            };
            let (mut iss, _) = issue(
                missing_code,
                "error",
                &format!(
                    "{}中的 Skill 目录丢失 / source skill directory is missing",
                    source_label
                ),
                Some("mark_skill_missing"),
            );
            add_extra(&mut iss, "skill_id", serde_json::json!(skill.id));
            add_extra(&mut iss, "skill_name", serde_json::json!(skill.name));
            add_extra(
                &mut iss,
                "community_path",
                serde_json::json!(source_path.to_string_lossy()),
            );
            issues.push(iss);
            continue;
        }

        // Check content hash drift
        if let Ok(source_hash) = content_hash::hash_dir(&source_path) {
            if let Some(ref stored_hash) = skill.content_hash {
                if stored_hash != &source_hash {
                    let (mut iss, _) = issue(
                        "source_hash_drift",
                        "info",
                        &format!(
                            "{}内容与数据库哈希不一致 / source content hash differs from DB",
                            source_label
                        ),
                        Some("update_content_hash"),
                    );
                    add_extra(&mut iss, "skill_id", serde_json::json!(skill.id));
                    add_extra(&mut iss, "stored_hash", serde_json::json!(stored_hash));
                    add_extra(&mut iss, "actual_hash", serde_json::json!(source_hash));
                    issues.push(iss);
                }
            }
        }

        // Check targets
        let targets = targets_repo.list_by_skill(&skill.id).unwrap_or_default();
        for target in &targets {
            target_paths.insert(path_safety::norm_path(&target.target_path));

            let target_path = Path::new(&target.target_path);
            let exists = target_path.exists()
                || target_path.is_symlink()
                || crate::platform::is_junction(target_path);

            if !exists {
                let (mut iss, _) = issue(
                    "missing_target_path",
                    "warning",
                    "数据库记录存在，但工具目录中的目标已丢失 / target path is missing",
                    Some("resync_target"),
                );
                add_extra(&mut iss, "skill_id", serde_json::json!(skill.id));
                add_extra(&mut iss, "tool", serde_json::json!(target.tool));
                add_extra(
                    &mut iss,
                    "target_path",
                    serde_json::json!(target.target_path),
                );
                issues.push(iss);
            }
        }
    }

    // Check for orphan dirs in community repo
    if community_base.is_dir() {
        if let Ok(entries) = std::fs::read_dir(&community_base) {
            let mut items: Vec<_> = entries.filter_map(|e| e.ok()).collect();
            items.sort_by_key(|e| e.file_name());
            for entry in items {
                let item = entry.path();
                if !item.is_dir() {
                    continue;
                }
                if entry.file_name() == ".snapshots" {
                    continue;
                }
                if !is_skill_dir(&item) {
                    continue;
                }
                let norm = path_safety::norm_path(&item);
                if !community_paths.contains(&norm) {
                    let (mut iss, _) = issue(
                        "community_orphan_dir",
                        "info",
                        "Community Repo 中存在未登记的 Skill / community skill is not registered",
                        Some("register_community_skill"),
                    );
                    add_extra(
                        &mut iss,
                        "community_path",
                        serde_json::json!(item.to_string_lossy()),
                    );
                    add_extra(
                        &mut iss,
                        "skill_name",
                        serde_json::json!(entry.file_name().to_string_lossy()),
                    );
                    issues.push(iss);
                }
            }
        }
    }

    // Check for duplicate content hashes
    let mut hash_groups: HashMap<String, Vec<&Skill>> = HashMap::new();
    for skill in &skills {
        if let Some(ref h) = skill.content_hash {
            hash_groups.entry(h.clone()).or_default().push(skill);
        }
    }
    for (hash, group) in &hash_groups {
        if group.len() > 1 {
            let (mut iss, _) = issue(
                "duplicate_content_hash",
                "info",
                "多个 Skill 内容完全相同 / duplicate skill content",
                None,
            );
            add_extra(
                &mut iss,
                "skill_ids",
                serde_json::json!(group.iter().map(|s| &s.id).collect::<Vec<_>>()),
            );
            add_extra(&mut iss, "content_hash", serde_json::json!(hash));
            issues.push(iss);
        }
    }

    let summary = summarize_issues(&issues);

    SyncHealthReport {
        community_repo: community_base.to_string_lossy().to_string(),
        skills_checked: skills.len(),
        issues,
        summary,
        generated_at: now_ms(),
    }
}

/// Run conservative repairs. Destructive deletes limited to DB records or managed targets.
pub fn repair_sync_health(db: &Database, dry_run: bool) -> RepairReport {
    use crate::repositories::SkillsRepository;

    let report = scan_sync_health(db);
    let mut operations: Vec<RepairOperation> = Vec::new();

    for issue_item in &report.issues {
        let action = match issue_item.repair_action.as_deref() {
            Some(a) => a,
            None => continue,
        };

        match action {
            "mark_skill_missing" => {
                let skill_id = issue_item
                    .extra
                    .get("skill_id")
                    .and_then(|v| v.as_str())
                    .unwrap_or("");
                operations.push(RepairOperation {
                    action: action.to_string(),
                    details: [("skill_id".to_string(), serde_json::json!(skill_id))]
                        .into_iter()
                        .collect(),
                });
                if !dry_run && !skill_id.is_empty() {
                    let repo = SkillsRepository::new(db);
                    if let Ok(Some(mut skill)) = repo.get_by_id(skill_id) {
                        skill.status = "missing".to_string();
                        skill.updated_at = now_ms();
                        let _ = repo.upsert(&mut skill);
                    }
                }
            }
            "update_content_hash" => {
                let skill_id = issue_item
                    .extra
                    .get("skill_id")
                    .and_then(|v| v.as_str())
                    .unwrap_or("");
                let actual_hash = issue_item
                    .extra
                    .get("actual_hash")
                    .and_then(|v| v.as_str())
                    .unwrap_or("");
                operations.push(RepairOperation {
                    action: action.to_string(),
                    details: [("skill_id".to_string(), serde_json::json!(skill_id))]
                        .into_iter()
                        .collect(),
                });
                if !dry_run && !skill_id.is_empty() {
                    let repo = SkillsRepository::new(db);
                    if let Ok(Some(mut skill)) = repo.get_by_id(skill_id) {
                        skill.content_hash = Some(actual_hash.to_string());
                        skill.updated_at = now_ms();
                        let _ = repo.upsert(&mut skill);
                    }
                }
            }
            _ => {
                operations.push(RepairOperation {
                    action: action.to_string(),
                    details: issue_item.extra.clone(),
                });
            }
        }
    }

    let after = if !dry_run {
        Some(scan_sync_health(db))
    } else {
        None
    };

    RepairReport {
        dry_run,
        operation_count: operations.len(),
        operations,
        before: report,
        after,
    }
}

/// Resolve the source path for a skill record.
fn resolve_skill_source_path(skill: &Skill, db: &Database) -> Result<std::path::PathBuf, String> {
    let community_path = &skill.community_path;
    if community_path.is_empty() {
        return Err("empty community path".to_string());
    }

    let path = std::path::PathBuf::from(community_path);
    let source_type = crate::repo::scanner::normalize_source_type(&skill.source_type);

    let base = if source_type == "custom" {
        resolve_custom_repo_path(db)
    } else {
        resolve_community_repo_path(db)
    };

    // Custom source type allows paths outside the repo
    if source_type == "custom" {
        return Ok(path);
    }

    if path_safety::is_path_within(&path, &base) {
        Ok(path)
    } else {
        Err(format!("source path escapes repo: {}", community_path))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_scan_sync_health_empty_db() {
        let db = Database::new_in_memory().unwrap();
        let report = scan_sync_health(&db);
        // No manually installed skills in a fresh DB
        assert_eq!(report.skills_checked, 0);
        // There should be no error-level issues for an empty DB
        assert!(
            report.issues.iter().all(|i| i.severity != "error"),
            "empty DB should not produce error-level issues"
        );
    }

    #[test]
    fn test_summarize_issues() {
        let issues = vec![
            HealthIssue {
                code: "a".to_string(),
                severity: "error".to_string(),
                message: "err".to_string(),
                repair_action: Some("fix".to_string()),
                extra: HashMap::new(),
            },
            HealthIssue {
                code: "b".to_string(),
                severity: "warning".to_string(),
                message: "warn".to_string(),
                repair_action: None,
                extra: HashMap::new(),
            },
            HealthIssue {
                code: "c".to_string(),
                severity: "info".to_string(),
                message: "info".to_string(),
                repair_action: Some("register".to_string()),
                extra: HashMap::new(),
            },
        ];
        let summary = summarize_issues(&issues);
        assert_eq!(summary.error, 1);
        assert_eq!(summary.warning, 1);
        assert_eq!(summary.info, 1);
        assert_eq!(summary.repairable, 2);
    }

    #[test]
    fn test_repair_sync_health_dry_run() {
        let db = Database::new_in_memory().unwrap();
        let report = repair_sync_health(&db, true);
        assert!(report.dry_run);
        assert!(report.after.is_none());
    }
}
